src/utils/PolyatomicDisplacementFunctionBase.C

```diff
@@ -0,0 +1,279 @@
+}
+
+/**
+ * The scattering cross section is computed using Lindhard's universal formula
+ * with f(xi) evaluated using the Thomas Fermi potential. Units are Angstrom**2 / eV
+ */
+Real
+PolyatomicDisplacementFunctionBase::scatteringCrossSection(unsigned int i,
+                                                           unsigned int j,
+                                                           Real energy,
+                                                           Real recoil_energy) const
+{
+  assert(i < _n_species);
+  assert(j < _n_species);
+
+  // get the current A & Z for projectile: i and target: j
+  Real Ai = _material->_element[i]._m;
+  Real Aj = _material->_element[j]._m;
+  Real Zi = _material->_element[i]._Z;
+  Real Zj = _material->_element[j]._Z;
+
+  // Z only appears as 1/3 power, so it's better to apply the 1/3 power here to get a sqrt
+  Real Z = std::sqrt(std::pow(Zi, 2.0 / 3.0) + std::pow(Zj, 2.0 / 3.0));
+  Real a = 0.8853 * _abohr / Z;
+
+  // compute El and Tm
+  Real Tm = _lambda[i][j] * energy;
+  Real El = 30.7514664 * Zi * Zj * Z * (Ai + Aj) / Aj;
+
+  // compute sqrt(t) and then the cross section using Lindhard's formula
+  Real sqrt_t = std::sqrt((energy * energy / El / El) * recoil_energy / Tm);
+  return 0.5 * universalF(sqrt_t) / recoil_energy / sqrt_t * M_PI * a * a;
+}
+
+Real
+PolyatomicDisplacementFunctionBase::universalF(Real xi) const
+{
+  Real lp = 1.309;
+  return lp * std::pow(xi, 1.0 / 3.0) *
+         std::pow(1.0 + std::pow(2.0 * lp * std::pow(xi, 4.0 / 3.0), 2.0 / 3.0), -1.5);
+}
+
+unsigned int
+PolyatomicDisplacementFunctionBase::findSpeciesIndex(unsigned int atomic_number,
+                                                     Real mass_number) const
+{
+  for (unsigned int j = 0; j < _n_species; ++j)
+    if (atomic_number == _material->_element[j]._Z &&
+        std::abs(mass_number - _material->_element[j]._m) < 1.0e-10)
+      return j;
+  return std::numeric_limits<unsigned int>::max();
+}
+
+Real
+PolyatomicDisplacementFunctionBase::displacementProbability(unsigned int k, Real energy) const
+{
+  return energy < _material->_element[k]._Edisp ? 0 : 1;
+}
+
+Real
+PolyatomicDisplacementFunctionBase::nonCaptureProbability(unsigned int i,
+                                                          unsigned int k,
+                                                          Real energy,
+                                                          Real recoil_energy) const
+{
+  // Parkin & Coulter, JNM 101, (1981)
+  return 1 -
+         displacementProbability(k, recoil_energy) * (energy - recoil_energy < _Ecap[i][k] ? 1 : 0);
+}
+
+unsigned int
+PolyatomicDisplacementFunctionBase::energyIndex(Real energy) const
+{
+  // quick check if energy < min_energy
+  if (energy < _energy_history[0])
+    return 0;
+
+  if (energy >= _energy_history.back())
+    return nEnergySteps() - 1;
+
+  // naive linear search; might have to fix if it's a performance drag
```

Comment:
is `_energy_history` sorted in any way?

Reply:
it must be sorted descending otherwise the comparison wouldn't make any sense.
You could use 

```
return std::distance(_energy_history.begin(), std::lower_bound(_energy_history.begin(), _energy_history.end(), energy));
```

here (I guess)